charts/k8ssandra/templates/_helpers.tpl

```diff
@@ -161,15 +161,54 @@ Set default num_tokens based on the server version
 */}}
 {{- define "k8ssandra.default_num_tokens" -}}
 {{- $datacenter := (index .Values.cassandra.datacenters 0) -}}
-    {{- if $datacenter.num_tokens }}
-      num_tokens: {{ $datacenter.num_tokens }}
-    {{- else }}
-    {{- if hasPrefix "3.11" .Values.cassandra.version }}
-      num_tokens: 256
-    {{- else }}
-      num_tokens: 16
-    {{- end }}
+{{- if .Release.IsInstall }}
+  {{- /*
+  If num_tokens is set we simply use that value.
+  */}}
+  {{- if $datacenter.num_tokens }}
+    {{- nindent 6 (print "num_tokens: " $datacenter.num_tokens) }}
+  {{- else }}
+  {{- /*
+  If num_tokens is not set we calculate the default value based on the C* version.
+  */}}
+  {{- if hasPrefix "3.11" .Values.cassandra.version }}
+    {{- nindent 6 (print "num_tokens: 256") }}
+  {{- else }}
+    {{- nindent 6 (print "num_tokens: 16") }}
+  {{- end }}
+{{- end }}
+{{- else }}
+  {{- /*
+  If upgrading and num_tokens is set simply use that value.
+  */}}
+  {{- if $datacenter.num_tokens }}
```

Comment:
So, to avoid the problem where the values are trying to change the num_tokens and they shouldn't, should we simply just not ever use the values on upgrade?  Always use what we find active in the currently deployed datacenter?

Reply:
Yeah, I think that sounds right to use whatever is set in the existing CassandraDatacenter, with a small exception. It is possible that we do not find an existing CassandraDatacenter in which case we should apply the value specified for `num_tokens` or apply the default log if the property is not set. I will update the PR accordingly.